pkg/synthetic/ticket.go

```diff
@@ -0,0 +1,186 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package synthetic generates fake data for testing.
+package synthetic
+
+import (
+	"fmt"
+	"math"
+	"math/rand"
+
+	structpb "github.com/golang/protobuf/ptypes/struct"
+	"open-match.dev/open-match/pkg/pb"
+	"open-match.dev/open-match/pkg/structs"
+)
+
+// Tickets generates artificial tickets with these properties:
+// mmr.rating = value 0 to 100, average 50, standard deviation of 15.
+// one of the role properties:
+//   role.dps = 1 with 40% chance
+//   role.support = 1 with 33% chance
+//   role.tank = 1 with 27% chance
+// mode.demo = 1 to 5 (inclusive) with decreasing percentage each mode.
+// region.europe-east1 = latency number from 0ms to ~300ms
+// region.europe-west1 = ^
+// region.europe-west2 = ^  (all of these latency numbers are really
+// region.europe-west3 = ^   bad and don't represent real world data)
+// region.europe-west4 = ^
+func Tickets(count int) []*pb.Ticket {
+	result := make([]*pb.Ticket, count)
+	for i := 0; i < count; i++ {
+		result[i] = CustomTicket(defaultGenerators...)
+	}
+	return result
+}
+
+var defaultGenerators = []GenFunc{
+	Normal("mmr.rating", 50, 15, 0, 100),
+	FieldChoice(map[string]int64{
+		"role.dps":     40,
+		"role.support": 33,
+		"role.tank":    27,
+	}),
+	ValueChoice("mode.demo", []*structpb.Value{
+		structs.Number(1),
+		structs.Number(2),
+		structs.Number(3),
+		structs.Number(4),
+		structs.Number(5),
+	}, []int64{
+		50, 30, 13, 5, 2,
+	}),
+	BadArtificalLatency(
+		"region.europe-east1",
+		"region.europe-west1",
+		"region.europe-west2",
+		"region.europe-west3",
+		"region.europe-west4",
+	),
+}
+
+// GenFunc defines a function which generates values for ticket properties.
+type GenFunc func(func(string, *structpb.Value))
+
+// CustomTicket creates a ticket using a custom list of generators.  If multiple
+// generators try to add a value to the same property name, CustomTicket will
+// panic.
+func CustomTicket(generators ...GenFunc) *pb.Ticket {
+	properties := structs.Struct{}
+	add := func(field string, value *structpb.Value) {
+		if _, ok := properties[field]; ok {
+			panic(fmt.Sprintf("Field %s populated with two differnet values.", field))
```

Comment:
Why not return errors from this library instead of panic? Fail and let the caller decide what to do?

Reply:
Correct usage of this library will never result in a panic.  This is more equivalent to passing a nil to something that panics because it wasn't expecting a nil value than it is to a network connection breaking.  Adding an error return means people need to check it, and they may assume that an errors happen in a reasonable case and that they need to properly handle them.